File: collcount.fs
\ Mary Bolling
\ Started: 4-28-2024
\ Last edited: 4-29-2024
\ CS 331, Assignment 7 Exercise B
\ Collatz Function, written in Forth

\ This function taken from the file word.fs by Glenn Chappell
: collmath { n -- c } \ Take variable n
    n 2 mod 0 = if \ Is n even?
        n 2 /
    else \ n must be odd
        n 3 * 1 +
    endif
;

: collcount ( n -- c )
    0 swap
    begin
        dup 1 = if \ You have reached 1, please follow the lighted strips to the exit
            drop exit
        else
            collmath \ Do some math
            swap 1 + swap \ Increment the number of times this has occurred
        endif
    again
;

File: collcount.pl
% Mary Bolling
% Started: 4-29-2024
% Last updated: 4-29-2024
% CS 331, Assignment 7 Exercise D
% Collatz Function, written in Prolog

% collcount/2
% End condition
collcount(1, 0).
% Do the math
collcount(N, C) :-
    N > 1,
    collmath(N, N1),
    % Check if end condition is met
    collcount(N1, P),
    C is P + 1.

% collmath/2
collmath(N, N1) :-
    % check even or odd
    Q is N mod 2,
    (Q = 0,
        % Even, set new N1
        N1 is N // 2
    ;Q = 1,
        % Odd, set new N1
        N1 is 3 * N + 1).

